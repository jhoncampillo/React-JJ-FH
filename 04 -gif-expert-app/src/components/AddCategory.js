import React, { useState } from "react";
import PropTypes from "prop-types";

//importo el estado
//puedo trae la props setCatecories con solo 'props'
export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState(""); //strig vacio

  //creo la funcion para capturar el evento e.target.value
  //se lo asigono al setInputValue
  //con el evento onChage cambio el contenido del input
  const handleInputChangue = (e) => {
    console.log(e.target.value);
    setInputValue(e.target.value);
  };

  //creo en evento enter
  const handleSubmit = (e) => {
    //quito el refresh del navegador
    e.preventDefault();
    console.log("Submit Echo");

    //avaluoo si son espcios
    if (inputValue.trim().length > 2) {
      //console.log(porsp.setCategories); asi seria con props
      //comono tengo acceso al inputValue c
      //adiciojo el inpul a los cata\s que ya hay
      setCategories((cats) => [...cats, inputValue]);
      setInputValue("");
    }
  };
  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChangue}
      ></input>
    </form>
  );
};
//Defino el props types requiered
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
