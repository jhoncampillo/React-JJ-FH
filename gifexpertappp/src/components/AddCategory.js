import React from "react";
import { useState } from "react";
import PropTypes from "prop-types";
//Traigo el parametro entre corchetes por el desetructuring
const AddCategory = ({ setCategories }) => {
  //creo el useState Para saber llo que la persona escribe en el input

  const [inputValue, setInputValue] = useState(""); //Seimpre es string Vacio

  //creo la funcion del onChangue
  const handleInputChangue = (e) => {
    setInputValue(e.target.value);

    //Log para pruebas
    console.log("handleInputChangue Llamado");
  };

  //CUANDO PULSA ENTER
  const handleSubmit = (e) => {
    //quita el refresh del navegador
    e.preventDefault();
    console.log("Submit OK- Handle Prueba 3 ", { inputValue });

    //valido espacios en blaco con trim
    if (inputValue.trim().length > 2) {
      //ADICIONO EL VALOR DEL INPUT AL LA <li></li>
      //Pondo de primera la ultima categoria solicitada
      setCategories((cats) => [inputValue, ...cats]);
      //reseteo en setinputValue
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      {/* paramtro para Prueba */}
      <p>{inputValue}</p>

      {/* <h2>Addcategory</h2> */}
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChangue}
        placeholder="Busca tu Personaje"
      />
    </form>
  );
};
//Defino el parametro setCategories como required
AddCategory.propTypes = {
  //defino una funcion como required
  setCategories: PropTypes.func.isRequired,
};

export default AddCategory;
