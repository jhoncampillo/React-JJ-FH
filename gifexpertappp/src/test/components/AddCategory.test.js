//Pruebas de AddCategory
import React from "react";
import { shallow } from "enzyme";
import AddCategory from "../../components/AddCategory";
//Autocompletado
import "@testing-library/jest-dom";

describe("Pruebas de  <AddCategory/>", () => {
  //Prueba 1 snapshopt

  //como el componente recibe una fucnion como required  temporalmente la vamos a crear

  //const setCategories = () => {}; la combio por:
  const setCategories = jest.fn();
  //asi  puedo saber como fue llamada - cuantas veces -por quien fue llamada

  //cambio   const Wrapper = shallow(<AddCategory setCategories={setCategories} />);
  //esta doble inicailizacion del warpper es para que la ayuda no desaparezca
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  //Ciclo de Vida
  beforeEach(() => {
    //limpio todas las simulaciones
    jest.clearAllMocks();
    //metro el warpper pero le cambio "const "por "let"
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("Debe de mostarse Correctamente ", () => {
    //defino el wrapper
    //const Wrapper = shallow(<AddCategory setCategories={setCategories} />);

    //esperado que sea igual a snapshot
    expect(wrapper).toMatchSnapshot();
  });

  //Prueba de evaluar cambion en la caja de Texto
  test("Evaluar  cambios en la caja de texto", () => {
    //Busco el Input en el jsx
    const input = wrapper.find("input");

    const value = "Hola Mundo";
    //simulo un cliclik con enzyne -  nose necesita Onclick - si no change
    //tengo que evaluar el evento.target.value que se evalua en el componente -
    //para eso creo una const arriba llamda value - liego creo el objeto
    //dentro del input,  {target:{value}}
    input.simulate("change", { target: { value } });

    expect(wrapper.find("p").text().trim()).toBe(value);
  });

  //Prueb del Submit

  test("No debe de postear la informacion onSubmit", () => {
    //busco en le wrapper el form y simulo el OnSubmit y le ,ando el preventDefault
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    //como tengo el inpul vacio la funcinio no se debe  llamar - se evalua eso
    //como el inputValue trae un valor .. tengo que menter el value dentro de
    //de un beforeEach - cilo de vida  en la parate superior
    expect(setCategories).not.toHaveBeenCalled();
  });

  //Otra Prueba para llamar setcategories y limpiar caja de texto
  //simula input changue- simular Submit - setCAtegories llamado
  //y el valor del input debe estar vacio
  test("Evaluar llamado setCategories, Simular Submit- InputChange -input vacio", () => {
    //valor para pruebas - mando el value
    const value = "Hola Mundo";
    // setCategories();
    //avaluao el Onchangue  evaluo el que el input no este vacio
    wrapper.find("input").simulate("change", { target: { value } });
    //simulo el submit - envio el value- e invovo la funcion setCategories
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    //evaluo llmados de la function
    expect(setCategories).toHaveBeenCalled();
    // que se llame 1 vez
    expect(setCategories).toHaveBeenCalledTimes(1);
    //Que se llame como una funcion
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

    //limpiar caja de texto
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
