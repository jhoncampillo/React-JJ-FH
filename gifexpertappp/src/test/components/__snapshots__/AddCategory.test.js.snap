// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de  <AddCategory/> Debe de mostarse Correctamente  1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Busca tu Personaje"
    type="text"
    value=""
  />
</form>
`;
